crates/bevy_core/src/time/time.rs

```diff
@@ -1,95 +1,264 @@
+            self.seconds_since_startup = self.elapsed_since_startup.as_secs_f32();
+            self.seconds_since_startup_f64 = self.elapsed_since_startup.as_secs_f64();
+        } else {
+            self.first_update = Some(instant);
+            let raw_delta = instant - self.startup;
+
+            self.raw_elapsed_since_startup += raw_delta;
+            self.raw_seconds_since_startup = self.raw_elapsed_since_startup.as_secs_f32();
+            self.raw_seconds_since_startup_f64 = self.raw_elapsed_since_startup.as_secs_f64();
+
+            // avoid any rounding errors when speed is 1:1
+            let delta = if self.relative_speed != 1.0 {
+                raw_delta.mul_f64(self.relative_speed)
+            } else {
+                raw_delta
+            };
+
+            self.elapsed_since_startup += delta;
+            self.seconds_since_startup = self.elapsed_since_startup.as_secs_f32();
+            self.seconds_since_startup_f64 = self.elapsed_since_startup.as_secs_f64();
         }
 
-        self.time_since_startup = instant - self.startup;
-        self.seconds_since_startup = self.time_since_startup.as_secs_f64();
         self.last_update = Some(instant);
     }
 
-    /// The delta between the current tick and last tick as a [`Duration`]
+    /// Returns the [`Instant`] the app was started.
+    #[inline]
+    pub fn startup(&self) -> Instant {
+        self.startup
+    }
+
+    /// Returns the [`Instant`] when [`Time::update`] was first called, if it exists.
+    #[inline]
+    pub fn first_update(&self) -> Option<Instant> {
+        self.first_update
+    }
+
+    /// Returns the [`Instant`] when [`Time::update`] was last called, if it exists.
+    #[inline]
+    pub fn last_update(&self) -> Option<Instant> {
+        self.last_update
+    }
+
+    /// Returns the rate that time advances relative to raw CPU time, as [`f32`].
+    ///
+    /// `1.0` by default.
+    #[inline]
+    pub fn relative_speed(&self) -> f32 {
+        self.relative_speed as f32
+    }
+
+    /// Returns the rate that time advances relative to raw CPU time, as [`f64`].
+    ///
+    /// `1.0` by default.
+    #[inline]
+    pub fn relative_speed_f64(&self) -> f64 {
+        self.relative_speed
+    }
+
+    /// Sets the rate that time advances relative to raw CPU time, given as [`f32`].
+    ///
+    /// # Panics
+    ///
+    /// Panics if `rate` negative or not finite.
+    #[inline]
+    pub fn set_relative_speed(&mut self, relative_speed: f32) {
+        assert!(relative_speed.is_sign_positive() && relative_speed.is_finite());
+        self.relative_speed = relative_speed as f64;
+    }
+
+    /// Sets the rate that time advances relative to raw CPU time, given as [`f64`].
+    ///
+    /// # Panics
+    ///
+    /// Panics if `rate` negative or not finite.
+    #[inline]
+    pub fn set_relative_speed_f64(&mut self, relative_speed: f64) {
+        assert!(relative_speed.is_sign_positive() && relative_speed.is_finite());
```

Comment:
```suggestion
        assert!(relative_speed.is_sign_positive() && relative_speed.is_finite(), "The relative speed has to be positive");
```
I would add an assert message here, it makes the error message a bit less scary for beginners.

Reply:
Same thing for the f32 variant

Reply:
Oh, good point.